File: frontend/src/types/tool.ts
/**
 * Tool Schema Definitions
 *
 * This file contains all TypeScript types and interfaces for defining Tools.
 */

import { SchemaEntity } from './base';
import { Resource } from './resource';

// --- Core Tool Interfaces ---

export interface ToolParameter extends SchemaEntity {
    required: boolean;
}

export interface ToolOutput extends SchemaEntity {
    required: boolean;
}

export interface ToolDefinition {
    id: string;
    name: string;
    description: string;
    category: string;
    functional_category?: string;
    domain_category?: string;
    tags?: string[];
    pipeline_info?: {
        can_start_pipeline: boolean;
        typical_next_tools: string[];
        pipeline_name: string;
    };
    ui_metadata?: {
        icon: string;
        color: string;
        difficulty: 'beginner' | 'intermediate' | 'advanced';
    };
    parameters: ToolParameter[];
    outputs: ToolOutput[];
    resource_dependencies: Resource[];
    examples?: Array<{
        description: string;
        input: Record<string, any>;
        output: Record<string, any>;
    }>;
}

// --- Tool Utility Functions ---

export function toolRequiresResources(tool: ToolDefinition): boolean {
    return tool.resource_dependencies.length > 0;
}

export function getToolResourceIds(tool: ToolDefinition): string[] {
    return tool.resource_dependencies.map(r => r.id);
}

export function getToolResourceConfig(tool: ToolDefinition, resourceId: string): Resource | undefined {
    return tool.resource_dependencies.find(r => r.id === resourceId);
} 

File: backend/schemas/tool.py
"""
Tool Schema Definitions

This module contains all Pydantic models and related utilities for defining
and managing Tools. Tools are the functional units that perform actions
within a hop.
"""

from __future__ import annotations
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

from .base import SchemaEntity
from .resource import Resource
from .tool_handler_schema import ToolExecutionHandler

# --- Core Tool Models ---

class ToolParameter(SchemaEntity):
    """
    Defines an input parameter for a tool. It extends SchemaEntity and adds
    a 'required' flag.
    """
    required: bool = Field(default=True)

class ToolOutput(SchemaEntity):
    """
    Defines an output field for a tool. It extends SchemaEntity and adds
    a 'required' flag.
    """
    required: bool = Field(default=True)

class ToolDefinition(BaseModel):
    """
    Represents the complete definition of a tool, including its parameters,
    outputs, and dependencies on external resources.
    """
    id: str
    name: str
    description: str
    category: str
    functional_category: Optional[str] = Field(default=None, description="Functional category (e.g., 'search_retrieve', 'extract_analyze')")
    domain_category: Optional[str] = Field(default=None, description="Domain category (e.g., 'academic_research', 'web_content')")
    tags: Optional[List[str]] = Field(default=None, description="Tags for the tool")
    pipeline_info: Optional[Dict[str, Any]] = Field(default=None, description="Pipeline information")
    ui_metadata: Optional[Dict[str, Any]] = Field(default=None, description="UI metadata")
    parameters: List[ToolParameter]
    outputs: List[ToolOutput]
    resource_dependencies: List[Resource] = Field(default_factory=list)
    
    # The execution_handler is not serialized and is attached at runtime.
    execution_handler: Optional[ToolExecutionHandler] = Field(default=None, exclude=True)

    # --- Utility Methods ---

    def requires_resources(self) -> bool:
        """Checks if the tool has any resource dependencies."""
        return len(self.resource_dependencies) > 0
    
    def get_resource_ids(self) -> List[str]:
        """Returns a list of IDs for all resources this tool depends on."""
        return [r.id for r in self.resource_dependencies]
    
    def get_resource_config(self, resource_id: str) -> Optional[Resource]:
        """Retrieves the definition for a specific resource dependency."""
        return next((r for r in self.resource_dependencies if r.id == resource_id), None) 